platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformEdgeCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,658 @@
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(color, "blue")
+                    from.attribute(texture, "unknown")
+                    to.attribute(color, "purple")
+                    to.attribute(texture, "rough")
+                }
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(color, "purple")
+                    to.attribute(color, "red")
+                }
+
+                // square -> triangle -> round
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(shape, "square")
+                    to.attribute(shape, "triangle")
+                }
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(shape, "triangle")
+                    to.attribute(shape, "round")
+                }
+
+                // blue -> yellow -> red
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(color, "blue")
+                    from.attribute(texture, "unknown")
+                    to.attribute(color, "yellow")
+                    to.attribute(texture, "smooth")
+                }
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(color, "yellow")
+                    to.attribute(color, "red")
+                }
+
+                // square -> flat -> round
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(shape, "square")
+                    to.attribute(shape, "flat")
+                }
+                registerTransform(BrokenTransform::class.java) {
+                    from.attribute(shape, "flat")
+                    to.attribute(shape, "round")
+                }
+            }
+
+            val forceResolution by tasks.registering {
+                inputs.files(configurations.getByName("resolveMe").incoming.artifactView {
+                    // After getting initial square-blue-liquid variant with liquid request, we request something red-round
+                    // There should be 2 separate transformation chains of equal length that produce this
+                    attributes.attribute(color, "red")
+                    attributes.attribute(shape, "round")
+                }.artifacts.artifactFiles)
+
+                doLast {
+                    inputs.files.files.forEach { println(it) }
+                }
+            }
+        """
+
+        expect:
+        fails "forceResolution"
+
+        failure.assertHasDescription("Could not determine the dependencies of task ':forceResolution'.")
+        failure.assertHasCause("Could not resolve all dependencies for configuration ':resolveMe'.")
+        failure.assertHasErrorOutput("""   > Found multiple transformation chains that produce a variant of 'root project :' with requested attributes:
```

Comment:
It is clear this error format can produce some huge walls of text. 

I left one suggestion earlier that I think could make this a little more consise.

Another issue I think with this format is that we duplicate transform steps. If a transform step appears in more than one transform chains, we repeat it.

For example:
```
- 'BrokenTransform':
  - Converts from attributes:
    - shape 'square'
  - To attributes:
    - shape 'triangle'
```
appears multiple times in the below error message.

A potential solution (just throwing out ideas) would be to perhaps number all participant transforms and print them out once:

```
Found multiple transformation chains that produce a matching artifact set.
Using the following registered artifact transforms:
  - (1) 'BrokenTransform'
    - Transforms color from 'blue' to 'red'
    - Transforms state from 'liquid' to 'gas'
    - Transforms from attr 'value'
  - (2) 'BrokenTransform'
    - Transforms color from 'foo' to 'bar'
    - Transforms state from 'asdfasdf' to 'aaaa'
    - Transforms from attr 'oi'
The following artifact sets can be constructed:
  - From source variant 'foo'
    - Transform Chain: (2) -> (1)
    - Transform Chain: (1) -> (2)
  - From source variant 'bar'
    - Transform Chain: (3) -> (2)
```

Reply:
The main point being is that in addition to long output text, it is difficult in the proposed error message to distinguish whether transforms are the same between two transform chains. In practice two transform chains use the same steps, but that is not immediately obvious in the output here

Reply:
I don't know.  My initial reaction is that if the error message needs a key to decode it, that's not good.

As I said before, I don't think being concise here is the goal.  I think it is quicker to understand the message when it is comprehensive.

I also think, with the current state of things (until #30784 and/or #30785), it would be **really** confusing to see a distinct (1) and (2) with identical everything...

I'd defer tweaking the message like this until those other changes can be considered.  This change takes the message from pretty long (but missing important information) to slightly longer (with all important information present), so I think it's a good step to deliver before worrying about condensing the message.